ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/storage/keyvalue/KeyValueStorageWorldStateStorage.java

```diff
@@ -12,17 +12,22 @@
  */
 package tech.pegasys.pantheon.ethereum.storage.keyvalue;
 
+import static java.nio.charset.StandardCharsets.UTF_8;
+
 import tech.pegasys.pantheon.ethereum.core.Hash;
 import tech.pegasys.pantheon.ethereum.trie.MerklePatriciaTrie;
 import tech.pegasys.pantheon.ethereum.worldstate.WorldStateStorage;
 import tech.pegasys.pantheon.services.kvstore.KeyValueStorage;
 import tech.pegasys.pantheon.util.bytes.Bytes32;
 import tech.pegasys.pantheon.util.bytes.BytesValue;
+import tech.pegasys.pantheon.util.bytes.BytesValues;
 
 import java.util.Optional;
 
 public class KeyValueStorageWorldStateStorage implements WorldStateStorage {
 
+  private static final BytesValue AVAILABILITY_PREFIX = BytesValue.of("available".getBytes(UTF_8));
```

Comment:
I was thinking we could handle this by flipping the approach and saving "provisional" state roots in a separate location.  For now, the only provisional root would be from fast sync.  When fast sync is done, we could then confirm that provisional state root and move it into regular storage.  The advantages of flipping the approach are: no need to migrate existing dbs and more efficient storage (no need to store extra data for every valid root).

Reply:
Switched to considering a state as available if the root node is present.  Fast sync delays committing the root node until the rest of the state has been downloaded.  Also realised that when the world state downloader is first started, rather than checking if the world state is empty, it can check if the world state it's after is available and skip downloading if it is (the empty world state is always available).